File: root/branches/3.0.0/modules/strategy/src/test/sample_data/inputs/send_event.rb
include_class "org.marketcetera.strategy.ruby.Strategy"
include_class "org.marketcetera.strategy.OutputType"
include_class "org.marketcetera.event.impl.TradeEventBuilder"
include_class "org.marketcetera.marketdata.DateUtils"
include_class "org.marketcetera.trade.Equity"
include_class "org.marketcetera.module.DataRequest"
include_class "java.lang.Long"
include_class "java.util.Date"
include_class "java.math.BigDecimal"

class SendEvent < Strategy
    def on_start
        @askCounter = 0
        dataRequests = Array.new
        dataRequests << DataRequest.new(get_urn, OutputType::ALL)
        @dataFlowID = create_data_flow true, dataRequests.to_java(DataRequest)
    end
    def on_stop
        tradeBuilder = TradeEventBuilder.equityTradeEvent
        tradeBuilder.withInstrument(Equity.new("METC")).withExchange("exchange")
        tradeBuilder.withPrice(BigDecimal::ONE).withSize(BigDecimal::TEN)
        tradeBuilder.withTradeDate(DateUtils.dateToString(Date.new))
        do_send_event tradeBuilder.create
        cancel_data_flow @dataFlowID
    end
    def on_other (data)
        do_send_event data
    end
    def on_callback(data)
        if(get_property("shouldRequestCEPData") != nil)
            do_cep_request
        else
            cancel_all_data_requests
        end
    end
    def on_ask(ask)
        @askCounter += 1
        set_property "askCount", Long.toString(@askCounter)
        set_property "ask", ask.toString
    end
    def do_send_event (event)
        if(get_property("eventOnlyTest") != nil)
            if(get_property("nilEvent") != nil)
                send_event nil
                return
            end
            send_event event
        else
            source = get_property "source"
            if(get_property("nilSource") != nil)
                send_event_to_cep event, nil  
                return
            end
            if(get_property("nilEvent") != nil)
                send_event_to_cep nil, source
                return
            end
            send_event_to_cep event, source
        end
    end
    def do_cep_request
        cepDataSource = get_property "source"
        if(cepDataSource != nil)
            statementString = get_property "statements"
            if(statementString != nil) 
                statements = statementString.split("#")
            else
                statements = nil
            end
            set_property("requestID", Long.toString(request_cep_data(statements.to_java(:string), cepDataSource)))
        end
    end
end

File: root/branches/3.0.0/modules/strategy/src/test/sample_data/inputs/SendEvent.java
import java.math.BigDecimal;

import org.marketcetera.core.event.AskEvent;
import org.marketcetera.core.event.Event;
import org.marketcetera.core.event.TradeEvent;
import org.marketcetera.core.event.impl.TradeEventBuilder;
import org.marketcetera.marketdata.DateUtils;
import org.marketcetera.core.module.DataFlowID;
import org.marketcetera.core.module.DataRequest;
import org.marketcetera.core.module.ModuleURN;
import org.marketcetera.strategy.OutputType;
import org.marketcetera.strategy.java.Strategy;
import org.marketcetera.core.trade.Equity;

import java.util.Date;

/* $License$ */

/**
 * Sample strategy which tests the ability to send events to CEP and event subscribers.
 *
 * @author <a href="mailto:colin@marketcetera.com">Colin DuPlantis</a>
 * @version $Id: SendEvent.java 16063 2012-01-31 18:21:55Z colin $
 * @since 1.0.0
 */
public class SendEvent
        extends Strategy
{
    private int askCounter = 0;
    private DataFlowID dataFlowID;
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onStart()
     */
    @Override
    public void onStart()
    {
        dataFlowID = createDataFlow(true,
                                    new DataRequest(getURN(),
                                                    OutputType.ALL));
    }
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onStart()
     */
    @Override
    public void onStop()
    {
        TradeEvent trade = TradeEventBuilder.equityTradeEvent().withInstrument(new Equity("METC"))
                                                               .withExchange("exchange")
                                                               .withPrice(BigDecimal.ONE)
                                                               .withSize(BigDecimal.TEN)
                                                               .withTradeDate(DateUtils.dateToString(new Date())).create();
        doSendEvent(trade);
        cancelDataFlow(dataFlowID);
    }
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onAsk(org.marketcetera.core.event.AskEvent)
     */
    @Override
    public void onAsk(AskEvent inAsk)
    {
        askCounter += 1;
        setProperty("askCount",
                    Integer.toString(askCounter));
        setProperty("ask",
                    inAsk.toString());
    }
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onCallback(java.lang.Object)
     */
    @Override
    public void onCallback(Object inData)
    {
        if(getProperty("shouldRequestCEPData") != null) {
            doCepRequest();
        } else {
            cancelAllDataRequests();
        }
    }
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onOther(java.lang.Object)
     */
    @Override
    public void onOther(Object inEvent)
    {
        doSendEvent((Event)inEvent);
    }
    /**
     * Sends the given event.
     *
     * @param inEvent an <code>Event</code> value
     */
    private void doSendEvent(Event inEvent)
    {
        if(getProperty("eventOnlyTest") != null) {
            if(getProperty("nilEvent") != null) {
                sendEvent(null);
                return;
            }
            sendEvent(inEvent);
        } else {
            String source = getProperty("source");
            if(getProperty("nilSource") != null) {
                sendEventToCEP(inEvent,
                               null);  
                return;
            }
            if(getProperty("nilEvent") != null) {
                sendEventToCEP(null,
                               source);
                return;
            }
            sendEventToCEP(inEvent,
                           source);
        }
    }
    /**
     * Sets up a CEP request.
     */
    private void doCepRequest()
    {
        String cepDataSource = getProperty("source");
        String statementString = null;
        if(cepDataSource != null) {
            statementString = getProperty("statements");
        }
        String[] statements;
        if(statementString != null) { 
            statements = statementString.split("#");
        } else { 
            statements = null;
        }
        setProperty("requestID",
                    Long.toString(requestCEPData(statements,
                                                 cepDataSource)));
    }
}
